vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_context.h

```diff
@@ -0,0 +1,272 @@
+  auto StartTask(Args&&... args) {
+    std::unique_ptr<T> instance(new T(this));
+    T* ptr = instance.get();
+
+    size_t key = std::hash<T*>()(ptr);
+    tasks_[key] = std::move(instance);
+
+    ptr->Start(std::forward<Args>(args)...);
+    auto result = ptr->result();
+
+    using CompleteType = typename decltype(result)::CompleteType;
+    auto on_complete =
+        base::BindOnce(&BATLedgerContext::OnTaskCompleted<CompleteType>,
+                       weak_factory_.GetWeakPtr(), key);
+
+    result.Listen(std::move(on_complete));
+    return result;
+  }
+
+  // The Log* functions return a LogStream used to log messages to the client.
+  // Log levels kError, kInfo, and kVerbose may be logged to disk by the client
+  // and should not contain any information that would result in a breach of
+  // security.
+  //
+  // Example:
+  //   context()->LogError(FROM_HERE) << "Something didn't work quite right";
+  //
+  // Since non-temporary objects are destroyed in FIFO order, be careful when
+  // binding a LogStream to an lvalue, as it could result in a surprising
+  // ordering of logging calls to the client.
+  //
+  //   LogStream stream1 = context()->LogInfo(FROM_HERE);
+  //   stream1 << "A";
+  //   LogStream stream2 = context()->LogInfo(FROM_HERE);
+  //   stream2 << "B";
+  //
+  // In the example above, |stream2| is destroyed first, and therefore sends its
+  // message to the client first.
+  LogStream Log(base::Location location, LogLevel log_level);
+  LogStream LogError(base::Location location);
+  LogStream LogInfo(base::Location location);
+  LogStream LogVerbose(base::Location location);
+  LogStream LogFull(base::Location location);
+
+  // Returns a const reference to a settings object of type T appropriate for
+  // the current ledger environment.
+  //
+  // Example:
+  //   struct MySettings {
+  //     static const MySettings kDevelopment;
+  //     static const MySettings kStaging;
+  //     static const MySettings kProduction;
+  //
+  //     int magic_number;
+  //   };
+  //
+  //   const MySettings MySettings::kDevelopment = {.magic_number = 1};
+  //   const MySettings MySettings::kStaging = {.magic_number = 2};
+  //   const MySettings MySettings::kProduction = {.magic_number = 3};
+  //
+  //   context()->GetSettings<MySettings>().magic_number;
+  template <typename T>
+  const T& GetSettings() {
+    switch (ledger::_environment) {
+      case mojom::Environment::DEVELOPMENT:
+        return T::kDevelopment;
+      case mojom::Environment::STAGING:
+        return T::kStaging;
+      case mojom::Environment::PRODUCTION:
+        return T::kProduction;
+    }
+  }
+
+ private:
+  template <typename T>
+  void OnTaskCompleted(size_t task_key, const T&) {
+    tasks_.erase(task_key);
+  }
+
+  LedgerClient* ledger_client_;
+  LedgerImpl* ledger_impl_ = nullptr;
+  std::unordered_map<size_t, std::unique_ptr<Component>> components_;
+  std::unordered_map<size_t, std::unique_ptr<Component>> tasks_;
```

Comment:
what would happen if the context dies before tasks are completed? Can't see any code canceling the tasks

Reply:
The task does not care who owns it. Generally, it will use a `WeakPtrFactory`, so any callbacks that it has created and registered will no-op. The associated `AsyncResult` for the task will be left in an always-pending state, and the AsyncResult's internal "state" data structure will be cleaned up once there are no more live instances of AsyncResult pointing to it.

Adding a "Cancel" method to tasks would be interesting (where presumably "Cancel" would be called by the context's destructor?), but it could result in subtle teardown ordering issues, and I don't think we need it for `bat-native-ledger`.

Reply:
I don't think there is any need for "Cancel", I just wanted to understand how a task lifetime is managed. If there is a weak pointer than someone has to invalidate the weak pointer on a proper sequence - just don't quite follow how this is supposed to work with this map

Reply:
In this model, the weak pointer would be invalidated by the `WeakPtrFactory` that is owned by the `Component` that is owned by the map that is owned by the BATLedgerContext. So a context and all of its components would live, and have to remain on, the same sequence. That was the idea anyway.

Reply:
> so is it supposed that a task running on another sequence can dereference that weak pointer?

No.

I think there might be some terminology confusion. (My fault!) "Task" here means: a ledger component that implements both a `Start` method and a `result() -> AsyncResult` method, and abstracts over a chain of async operations. We have these "chains of async ops" all over the place in ledger. I'm trying to think of a name that will distinguish this concept from a Chromium "task", which is some callback that's been posted to a `TaskRunner`, but I haven't been able to come up with another name yet. Let's just call it an "AsyncOp" for now. All AsyncOps that are registered with the ledger context would have their WeakPtrs deref'd on the same sequence as the context. The management of these "in progress" AsyncOps is separate from whether we're doing any multithreading.

Ledger doesn't do any multithreading at the moment, but in my imagination it would work something like:

- Some ledger component (perhaps an AsyncOp) wants to do something in the background, so it creates an `AsyncResult<T>::Resolver` to represent the result of that something.
- It passes that resolver, along with any additional (thread-safe) data to the other sequence through a `TaskRunner`.
- Work is done on the other sequence, when finished, it calls `resolver.Complete(...)`.
- Internally, that posts a task to the originating sequence to update the async result data structure and post additional tasks to any sequences that have registered a listener for the async result.

Hope that helps!